### ISSUE TITLE
[codegen] ICE on >=homestead spuriousDragon<= in ReturnInfo constructor because `haveReturndatacopy` is false

### ISSUE THREAD

**Issue by bshastry** [ethereum/solidity#9815]
## Description

```
pragma experimental ABIEncoderV2;
contract C {
  function f() public returns(bool[][4] memory) {
    this.f();
  }
}
```

throws

https://github.com/ethereum/solidity/blob/f964966090f7e3571e20144fd4aa0b6391aea26e/libsolidity/codegen/ReturnInfo.cpp#L46

## Environment

- Compiler version: latest develop
- Target EVM version (as per compiler settings): Homestead, TangerineWhistle, SpuriousDragon

## Steps to Reproduce

```
$ solc --evm-version homestead test.sol
```

**Comment by cameel** [comment]
It looks like `FunctionType::returnParameterTypesWithoutDynamicTypes()` is returning a parameter for which `isDynamicallyEncoded()` is true.

This looks straightforward to debug and fix so I'm going to take it. I could use a short break from the other ABI encoder bug where I have to play whack-a-mole with failing test cases :)

**Comment by cameel** [comment]
Fix: #9819.

Turns out that `returnParameterTypesWithoutDynamicTypes()` was checking the type using `isDynamicallySized()`, which is not recursive and only checks the outer array. Changing it to `isDynamicallyEncoded()` fixes the problem.

I also tried to review other uses of `isDynamicallySized()` in our code to make sure there are no similar errors. It's used a lot though and all uses in `Types` and `TypeChecker` were fine so I gave up on that in the end.
